ServerCore/Pages/Responses/CreateBulkMulti.cshtml.cs

```diff
@@ -0,0 +1,199 @@
+                {
+                    Puzzle puzzle = await (from Puzzle p in _context.Puzzles
+                                           where p.Name == puzzleName && p.EventID == this.Event.ID
+                                           select p).FirstOrDefaultAsync();
+
+                    if (puzzle == null)
+                    {
+                        ModelState.AddModelError("PuzzleName", $"Puzzle '{puzzleName}' not found");
+                        break;
+                    }
+
+                    if (EventRole == EventRole.author && !await UserEventHelper.IsAuthorOfPuzzle(_context, puzzle, LoggedInUser))
+                    {
+                        ModelState.AddModelError("PuzzleName", $"Not an author of puzzle '{puzzleName}'");
+                        break;
+                    }
+
+                    puzzleId = puzzle.ID;
+
+                    puzzleTitleLookup[puzzleName] = puzzleId;
+                    submissionsLookup[puzzleName] = new HashSet<string>();
+
+                    if (DeleteExisting)
+                    {
+                        Response[] responsesToRemove = await (from Response r in _context.Responses
+                                                              where r.PuzzleID == puzzleId
+                                                              select r).ToArrayAsync();
+                        _context.Responses.RemoveRange(responsesToRemove);
+                    }
+                }
+
+                // TODO probably clearer ways to validate but I honestly do not understand how validation works
+                if (puzzleName == null)
+                {
+                    if (submittedText != null)
+                    {
+                        ModelState.AddModelError("SubmittedText", "Unmatched Submission without Puzzle");
+                    }
+                    if (responseText != null)
+                    {
+                        ModelState.AddModelError("ResponseText", "Unmatched Response without Puzzle");
+                    }
+                    if (isSolution != null)
+                    {
+                        ModelState.AddModelError("IsSolution", "Unmatched IsSolution without Puzzle");
+                    }
+                    if (note != null)
+                    {
+                        ModelState.AddModelError("Note", "Unmatched Note without Puzzle");
+                    }
+
+                    // we're done
+                    break;
+                }
+
+                if (submittedText == null)
+                {
+                    ModelState.AddModelError("SubmittedText", "Unmatched Puzzle without Submission");
+                    break;
+                }
+
+                string submittedTextFormatted = ServerCore.DataModel.Response.FormatSubmission(submittedText);
+
+                // Ensure that the submission text is unique for this puzzle.
+                async Task<bool> isSubmittedTextUnique(string text)
```

Comment:
I wonder if it would be more efficient to select all responses for the puzzle the first time we see it and add them to submissionsLookup. (A DB constraint would work too, but I assume you're trying to be more user friendly here)

Reply:
I think you're right about this; I'll change it.